### PR TITLE
Implement OAUTHBEARER server

### LINKED ISSUE #16: Add OAUTHBEARER server
We only have the client side part currently.

### PR DESCRIPTION
**Pull request by foxcpp**
Closes #16.
Ref. https://github.com/foxcpp/maddy/issues/169

### CODE REVIEW

File: oauthbearer.go [hunk 1]
```diff
@@ -61,3 +64,121 @@ func (a *oauthBearerClient) Next(challenge []byte) ([]byte, error) {
...
+	// Per RFC, we cannot just send an error, we need to return JSON-structured
+	// value as a challenge and then after getting dummy response from the
+	// client stop the exchange.
+	if a.fail != nil {
+		if len(response) != 1 && response[0] != 0x01 {
+			return nil, true, errors.New("unexpected response")
+		}
+		return nil, true, a.fail
+	}
+	fail := func(descr string) ([]byte, bool, error) {
```
Comment (emersion) [inline oauthbearer.go:86]: Can we avoid nesting this function by making it an `oauthBearerServer` method?

File: oauthbearer.go [hunk 2]
```diff
@@ -61,3 +64,121 @@ func (a *oauthBearerClient) Next(challenge []byte) ([]byte, error) {
...
+		if err != nil {
+			panic(err) // wtf
+		}
+		a.fail = errors.New(descr)
+		return blob, false, nil
+	}
+
+	if a.done {
+		err = ErrUnexpectedClientResponse
+		return
```
Comment (emersion) [inline oauthbearer.go:100]: These should be avoided

https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

File: oauthbearer.go (hunk 1 above)
Comment (emersion) [reply oauthbearer.go:86]: (Also we could name it `invalidRequest` instead of `fail`)

File: oauthbearer.go (hunk 2 above)
Comment (emersion) [reply oauthbearer.go:100]: (The rest of the codebase has the wrong style because it's old)

Comment (foxcpp) [reply oauthbearer.go:100]: You probably want to clean plain.go too then :)
https://github.com/emersion/go-sasl/blob/6fe53deab1f6dadb88217928f90c11179a5885df/plain.go#L47-L48

File: oauthbearer.go (hunk 1 above)
Comment (foxcpp) [reply oauthbearer.go:86]: Sure.

File: oauthbearer.go [hunk 3]
```diff
@@ -61,3 +64,121 @@ func (a *oauthBearerClient) Next(challenge []byte) ([]byte, error) {
...
+		case "host":
+			opts.Host = string(pParts[1])
+		case "port":
+			port, err := strconv.ParseUint(string(pParts[1]), 10, 16)
+			if err != nil {
+				return fail("Invalid response, malformed 'port' value")
+			}
+			opts.Port = int(port)
+		case "auth":
+			const prefix = "bearer "
```
Comment (emersion) [inline oauthbearer.go:157]: Should we support other token types?

Comment (emersion) [reply oauthbearer.go:164]: Doesn't seem like it.

Comment (foxcpp) [reply oauthbearer.go:164]: RFC 7628 describes OAUTHBEARER mechanism as one using Bearer tokens. https://tools.ietf.org/html/rfc7628#section-3

No support for other token types is defined (besides OAUTH10B but this is a different mechanism).

**Pull request merged**